ddtrace/tracer/tracer.go

```diff
@@ -306,16 +316,50 @@ func (t *tracer) worker(tick <-chan time.Time) {
 	}
 }
 
-func (t *tracer) pushTrace(trace []*span) {
+// traceInfo contains span list and sampling decision of the trace
+// needed to run single span sampling.
+type traceInfo struct {
+	spans            []*span
+	samplingDecision samplingDecision
+}
+
+// processTraceInfo runs single span sampling on the spans pushed to the worker thread.
+// Sampling spans on the worker thread keeps any associated latency off the application thread.
+func (t *tracer) processTraceInfo(info *traceInfo) {
+	if info.samplingDecision == decisionKeep {
+		return
+	}
+	if !t.rulesSampling.HasSpanRules() {
+		info.spans = nil
+		return
+	}
+	// if trace sampling decision is drop, we still want to send single spans
+	// unless there are no single span sampling rules defined
+	var kept []*span
+	for _, span := range info.spans {
+		if t.rulesSampling.SampleSpan(span) {
+			kept = append(kept, span)
+		}
+	}
+	atomic.AddUint64(&t.droppedP0Spans, uint64(len(info.spans)-len(kept)))
+	info.spans = kept
+	if len(kept) == 0 {
+		atomic.AddUint64(&t.droppedP0Traces, 1)
+		return // no spans matched the rules and were sampled
+	}
+	atomic.AddUint64(&t.partialTraces, 1)
+}
+
+func (t *tracer) pushTraceInfo(info *traceInfo) {
```

Comment:
Despite you having changed the signature of this function by adding a new argument. I would not change the name. The "info" suffix is not needed imho.